### PR TITLE
Fix unmarshaling when additionalProperties is an object or true

### LINKED ISSUE #5: Unmarshaler for `additionalProperties` with object value incorrect?
If I'm not mistaken, the unmarshaler for `additionalProperties` with an object value is incorrect.

The [unit test](https://github.com/grktsh/python-oas/blob/e6cc018a6f5150ff5eaca1a44adabfc818d5be55/tests/schema/test_unmarshalers.py#L98-L122) tests the following value assigned to `additionalProperties`:

```json
{
  "additionalProperties": {
    "x": {
      "type": "string"
    }
  }
}
```

The [JSON Schema Draft 4](https://tools.ietf.org/html/draft-fge-json-schema-validation-00#section-5.4.4.1) states:

> The value of "additionalProperties" MUST be a boolean or an object. If it is an object, it MUST also be a valid JSON Schema.

An [example from Draft 7](https://json-schema.org/understanding-json-schema/reference/object.html) (I didn't find one for Draft 4, but I think there is no difference in this case) looks like this:

```json
{
  "additionalProperties": { "type": "string" }
}
```

I think the problem is that the unmarshaler treats `additionalProperties` the same way as `properties` when `additionalProperties` is an `object`/`dict`:

https://github.com/grktsh/python-oas/blob/e6cc018a6f5150ff5eaca1a44adabfc818d5be55/src/oas/schema/unmarshalers.py#L75-L76

Instead, I think `properties` should be unmarshaled as it's done now, but any remaining instance properties should be unmarshaled using the `additionalProperties` subschema, if present, using [`SchemaUnmarshaler::_unmarshal`](https://github.com/grktsh/python-oas/blob/e6cc018a6f5150ff5eaca1a44adabfc818d5be55/src/oas/schema/unmarshalers.py#L22). The results of both unmarshalers should then be merged. Do you agree?

### PR DESCRIPTION
**Pull request by sisp**
Fixes #5 and #7.

### CODE REVIEW

File: src/oas/schema/unmarshalers.py [hunk 1]
```diff
@@ -85,6 +82,17 @@ def _unmarshal_object(self, instance, schema):
                 except KeyError:
                     continue
             result[name] = self._unmarshal(value, sub_schema)
+
+        additional_properties = schema.get('additionalProperties')
```
Comment (sisp) [inline src/oas/schema/unmarshalers.py]: Based on https://github.com/grktsh/python-oas/issues/5#issuecomment-488287591, should this line rather be as follows?

```python
additional_properties = schema.get('additionalProperties', True)
```

Comment (grktsh) [reply src/oas/schema/unmarshalers.py]: LGTM

**Comment by sisp** [comment]
`additionalProperties` now defaults to `true` and I've extended the tests accordingly.

File: tests/schema/test_unmarshalers.py [hunk 2]
```diff
@@ -98,14 +98,19 @@ def test_unmarshal_object():
...
+        (None, {'type': 'string'}, {'x': 'foo', 'y': 'bar'}),
+        ({'x': {'type': 'string'}}, None, {'x': 'foo', 'y': 'bar'}),
+        ({'x': {'type': 'string'}}, True, {'x': 'foo', 'y': 'bar'}),
+        ({'x': {'type': 'string'}}, False, {'x': 'foo'}),
+        ({'x': {'type': 'string'}}, {}, {'x': 'foo', 'y': 'bar'}),
+        (
+            {'x': {'type': 'string'}},
+            {'type': 'string'},
+            {'x': 'foo', 'y': 'bar'},
+        ),
```
[the remark this answers is not in the document]
Comment (sisp) [reply tests/schema/test_unmarshalers.py]: You're looking to make the tests more robust, right? I think it's a good idea.

Isn't this test invalid?

```python
(None, {'type': 'string', 'format': 'date'}, {'x': 'foo', 'y': datetime.date(2020, 1, 2)}),
```

`{'x': 'foo'}` cannot be unmarshaled using `{'type': 'string', 'format': 'date'}`.

[the remark this answers is not in the document]
Comment (grktsh) [reply tests/schema/test_unmarshalers.py]: > You're looking to make the tests more robust, right?

Yes, test data for dict additionalProperties don't represent that they will be unmarshaled.

> Isn't this test invalid?

Yes... Is this OK? Feel free to write your own tests!

```python
(None, {'type': 'string', 'format': 'date'}, {'x': datetime.date(2019,  1, 2), 'y': datetime.date(2020, 1, 2)}),
({'x': {'type': 'string'}}, None, {'x': '2019-01-02', 'y': '2020-01-02'}),
...
({'x': {'type': 'string'}}, {'type': 'string', 'format': 'date'}, {'x': '2019-01-02', 'y': datetime.date(2020, 1, 2)}),
...
instance = {'x': '2019-01-02'}
```

[the remark this answers is not in the document]
Comment (sisp) [reply tests/schema/test_unmarshalers.py]: What do you think about this?

```python
@pytest.mark.parametrize(
    'properties,additional_properties,instance,expected',
    [
        (
            None,
            None,
            {'x': 'foo', 'y': '2020-01-02'},
            {'x': 'foo', 'y': '2020-01-02'},
        ),
        (
            None,
            True,
            {'x': 'foo', 'y': '2020-01-02'},
            {'x': 'foo', 'y': '2020-01-02'},
        ),
        (
            None,
            {},
            {'x': 'foo', 'y': '2020-01-02'},
            {'x': 'foo', 'y': '2020-01-02'},
        ),
        (
            None,
            {'type': 'string', 'format': 'date'},
            {'y': '2020-01-02'},
            {'y': datetime.date(2020, 1, 2)},
        ),
        (
            {'x': {'type': 'string'}},
            None,
            {'x': 'foo', 'y': '2020-01-02'},
            {'x': 'foo', 'y': '2020-01-02'},
        ),
        (
            {'x': {'type': 'string'}},
            True,
            {'x': 'foo', 'y': '2020-01-02'},
            {'x': 'foo', 'y': '2020-01-02'},
        ),
        ({'x': {'type': 'string'}}, False, {'x': 'foo'}, {'x': 'foo'}),
        (
            {'x': {'type': 'string'}},
            {'type': 'string', 'format': 'date'},
            {'x': 'foo', 'y': '2020-01-02'},
            {'x': 'foo', 'y': datetime.date(2020, 1, 2)},
        ),
    ],
)
def test_unmarshal_object_properties_and_additional_properties(
    properties, additional_properties, instance, expected
):
    schema = {'type': 'object'}
    if properties is not None:
        schema['properties'] = properties
    if additional_properties is not None:
        schema['additionalProperties'] = additional_properties

    unmarshaled = SchemaUnmarshaler().unmarshal(instance, schema)
    assert unmarshaled == expected
```

[the remark this answers is not in the document]
Comment (grktsh) [reply tests/schema/test_unmarshalers.py]: LGTM

**Comment by sisp** [comment]
I've integrated the changes according to our discussion.

**Comment by grktsh** [comment]
Thanks!

**Comment by sisp** [comment]
Thanks for the discussions and your feedback!

**Pull request merged**